### PR TITLE
SMCCT-862 add ability to reset and change password

### PR DESCRIPTION
**Pull request by csampsonza**

### CODE REVIEW

File: packages/happn-3/lib/services/protocol/service.js [hunk 1]
```diff
@@ -223,20 +212,48 @@ function __processSinglePath(transformed, callback) {
       });
     }
 
+    if (authorized.request.action === 'remove')
+      return this.happn.services.data.processRemove(authorized, (e, publication) => {
+        if (e) return callback(e);
+        if (publication.request.options && publication.request.options.noPublish)
```
Comment (southbite) [inline packages/happn-3/lib/services/protocol/service.js:218]: may as well tidy this up with: if ( publication.request?.options?.noPublish)

File: packages/happn-3/lib/services/security/service.js [hunk 2]
```diff
@@ -744,6 +746,39 @@ module.exports = class SecurityService extends require('events').EventEmitter {
     return ttl || 0; // Infinity turns to null over the wire, 0 can be 0
   }
 
+  __resetPassword(authorized, callback) {
```
Comment (southbite) [inline packages/happn-3/lib/services/security/service.js:749]: may as well use #resetPassword instead of __resetPassword

File: packages/happn-3/lib/services/security/service.js [hunk 3]
```diff
@@ -744,6 +746,39 @@ module.exports = class SecurityService extends require('events').EventEmitter {
...
+        .catch((e) => {
+          error = e;
+        })
+        .finally(() => {
+          callback(error, authorized);
+        });
+    });
+  }
+
+  __changePassword(authorized, callback) {
```
Comment (southbite) [inline packages/happn-3/lib/services/security/service.js:765]: #changePassword instead of __changePassword

File: packages/happn-3/test/__fixtures/test/integration/security/authentication/workingAuth.js [hunk 4]
```diff
@@ -7,25 +9,32 @@ module.exports = class TestAuthProvider extends require('../../../../../..').pro
...
+  async providerTokenLogin(credentials, decodedToken, sessionId) {
+    if (decodedToken.username === 'secondTestuser@somewhere.com' && credentials.token != null) {
+      let user = { username: 'secondTestuser@somewhere.com', groups: [] };
       return this.loginOK(credentials, user, sessionId);
     }
     return this.loginFailed(credentials.username, 'Invalid credentials');
   }
 
+  async providerResetPassword(credentials) {
+    throw new Error(`Works !! Password reset ${credentials.username}`);
```
Comment (southbite) [inline packages/happn-3/test/__fixtures/test/integration/security/authentication/workingAuth.js:32]: I assume this is so you can test both the bubbling of an error and ensure the plugin method gets reached?

File: packages/happn-3/lib/providers/security-happn-auth-provider.js [hunk 5]
```diff
@@ -54,4 +54,20 @@ module.exports = class HappnAuthProvider extends SecurityBaseAuthProvider {
     }
     return this.loginOK(credentials, user, sessionId);
   }
+
+  async providerChangePassword(credentials, passwordDetails) {
+    const user = await this.securityFacade.users.getUser(credentials.username);
+    if (user == null) {
+      return this.accessDenied(`bad username: ${credentials.username}`);
+    }
+    if (passwordDetails?.oldPassword && passwordDetails?.newPassword) {
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-happn-auth-provider.js:63]: May want to exit early here, so:
```suggestion
if (!passwordDetails?.oldPassword || !passwordDetails?.newPassword) {
   return this.loginFailed(credentials.username, 'Invalid parameters: oldPassword and newPassword required')
}
const hash = await this.securityFacade.users.getPasswordHash(credentials.username);
if (!(await this.securityFacade.security.matchPassword(passwordDetails.oldPassword, hash))) {
        return this.accessDenied(`Invalid old password`);
}
user.password = passwordDetails.newPassword;
return this.securityFacade.users.upsertUser(user);
```

File: packages/happn-3/lib/services/security/service.js [hunk 6]
```diff
@@ -50,6 +50,8 @@ module.exports = class SecurityService extends require('events').EventEmitter {
     this.login = util.maybePromisify(this.login);
     this.matchPassword = util.maybePromisify(this.matchPassword);
     this.verifyAuthenticationDigest = util.maybePromisify(this.verifyAuthenticationDigest);
+    this.resetPassword = util.maybePromisify(this.__resetPassword);
```
Comment (southbite) [inline packages/happn-3/lib/services/security/service.js:53]: use this.#resetPassword and this.#changePassword

**Review by southbite — CHANGES_REQUESTED** [review]

File: packages/happn-3/lib/services/protocol/service.js (hunk 1 above)
Comment (csampsonza) [reply packages/happn-3/lib/services/protocol/service.js:218]: Which version of node are we targeting? Guessing > 16

File: packages/happn-3/test/__fixtures/test/integration/security/authentication/workingAuth.js (hunk 4 above)
Comment (csampsonza) [reply packages/happn-3/test/__fixtures/test/integration/security/authentication/workingAuth.js:32]: Checking we get the username too. This in a auth provider will probably just run with no errors on success.

File: packages/happn-3/lib/services/protocol/service.js (hunk 1 above)
Comment (csampsonza) [reply packages/happn-3/lib/services/protocol/service.js:217]: Fixed.

File: packages/happn-3/lib/client.js [hunk 7]
```diff
@@ -940,7 +940,10 @@
   };
 
   HappnClient.prototype.__performSystemRequest = function (action, data, options, callback) {
-    var message = {
+    if (typeof callback !== 'function') {
```
Comment (csampsonza) [inline packages/happn-3/lib/client.js:943]: @southbite Please review this.  I think if we call perform system request with wrong arguments.  Callback isn't a function.  Think it's safe to fail early in this case.

Comment (southbite) [reply packages/happn-3/lib/client.js:943]: Is this necessary? Considering __performSystemRequest is internal we should perhaps expect it to always be called with the correct arguments?

File: packages/happn-3/lib/client.js [hunk 8]
```diff
@@ -538,7 +538,7 @@
     };
   };
 
-  HappnClient.prototype.getScript = function (url, callback) {
+  HappnClient.prototype.getScript = utils.maybePromisify(function (url, callback) {
```
Comment (southbite) [inline packages/happn-3/lib/client.js:541]: Is there a reason these were wrapped? they are actually internal methods (so not part of the client API per se) - my next job is going to be rationalizing this client code - as it is very old...

File: packages/happn-3/package.json [hunk 9]
```diff
@@ -63,6 +63,7 @@
     "happn-db-provider-mongo": "1.1.8",
     "happn-random-activity-generator": "0.2.1",
     "jsprim": "^2.0.0",
+    "mocha": "^10.2.0",
```
Comment (southbite) [inline packages/happn-3/package.json:66]: why is this here? mocha should be available in happn-commons-test (which is also required)

File: packages/happn-3/package.json [hunk 10]
```diff
@@ -80,5 +81,8 @@
     ]
   },
   "protocol": "4",
-  "readmeFilename": "README.md"
+  "readmeFilename": "README.md",
+  "volta": {
```
Comment (southbite) [inline packages/happn-3/package.json:85]: volta looks like it could change the behaviour of command line tools for other folks, are we ok with this being added to package.json?

File: packages/happn-3/lib/client.js (hunk 7 above)
Comment (csampsonza) [reply packages/happn-3/lib/client.js]: Yes, because when calling the client we have no idea how many arguments there might be.  e.g. changePassword with no arguments, means data will be in callback.  We could check every incoming method before it's wrapped, but that is alot more code.

File: packages/happn-3/package.json (hunk 10 above)
Comment (csampsonza) [reply packages/happn-3/package.json]: Removed

File: packages/happn-3/lib/client.js (hunk 8 above)
Comment (csampsonza) [reply packages/happn-3/lib/client.js]: I needed to wrap revoke-token.  so thought do all functions on the prototype that take a callback

**Review by southbite — APPROVED** [review]
